core/domain/classifier_services.py

```diff
@@ -0,0 +1,338 @@
+# coding: utf-8
+#
+# Copyright 2015 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import operator
+
+import numpy
+
+
+class StringClassifier(object):
+    """Handles math-y internals for classifying strings.
+    https://en.wikipedia.org/wiki/Latent_Dirichlet_allocation"""
+
+    # Learning rates
+    _DEFAULT_ALPHA = 0.1
+    _DEFAULT_BETA = 0.001
+
+    # Default label
+    # Captures parts of docs that don't help with prediction
+    _DEFAULT_LABEL = '_default'
+
+    def __init__(self):
+        """
+        Hieroglyph decoder:
+        b - boolean
+        c - count
+        p - position
+        l - label (id)
+        w - word (id)
+        d - doc (id)
+
+        Internal (private) model representation:
+        _w_dp - word given doc and count
+            (list of word ids which represent a doc)
+        _b_dl - boolean given doc and label
+            (boolean of whether a doc has a label)
+        _l_dp - label given doc and count
+            (assigned label of each word instance in a doc)
+        _c_dl - count given doc and label
+            (count of each label per doc)
+        _c_lw - count given label and word
+            (count of labels per word)
```

Comment:
This could be more specific. Is it this?

    total number of times a label gets assigned to an instance of a specific word, aggregated over all docs

Hard to figure out just by looking at this file -- there needs to be a precise definition somewhere.